config/queue_test.go

```diff
@@ -0,0 +1,686 @@
+				go func() {
+					defer wg.Done()
+					for {
+						select {
+						case tg := <-ch:
+							got = append(got, tg)
+							if len(got) == 3 {
+								cancel()
+							}
+						case <-ctx.Done():
+							cancel()
+							return
+						}
+					}
+				}()
+
+				return ctx, ch, func() []*configpb.TestGroup {
+					wg.Wait()
+					return got
+				}
+			},
+			freq: time.Microsecond,
+			want: []*configpb.TestGroup{
+				{
+					Name: "hi",
+				},
+				{
+					Name: "hi",
+				},
+				{
+					Name: "hi",
+				},
+			},
+		},
+		{
+			name: "multi",
+			q: func() *TestGroupQueue {
+				var q TestGroupQueue
+				q.Init([]*configpb.TestGroup{
+					{
+						Name: "hi",
+					},
+					{
+						Name: "there",
+					},
+				}, time.Now())
+				return &q
+			}(),
+			receivers: func(ctx context.Context) (context.Context, chan<- *configpb.TestGroup, func() []*configpb.TestGroup) {
+				ch := make(chan *configpb.TestGroup)
+				var wg sync.WaitGroup
+				wg.Add(1)
+				var got []*configpb.TestGroup
+				ctx, cancel := context.WithCancel(ctx)
+				go func() {
+					defer wg.Done()
+					for {
+						select {
+						case tg := <-ch:
+							got = append(got, tg)
+							if len(got) == 2 {
+								cancel()
+							}
+						case <-ctx.Done():
+							cancel()
+							return
+						}
+					}
+				}()
+
+				return ctx, ch, func() []*configpb.TestGroup {
+					wg.Wait()
+					return got
+				}
+			},
+			want: []*configpb.TestGroup{
+				{
+					Name: "hi",
+				},
+				{
+					Name: "there",
+				},
+			},
+		},
+		{
+			name: "multi loop",
+			q: func() *TestGroupQueue {
+				var q TestGroupQueue
+				q.Init([]*configpb.TestGroup{
+					{
+						Name: "hi",
+					},
+					{
+						Name: "there",
+					},
+				}, time.Now())
+				return &q
+			}(),
+			receivers: func(ctx context.Context) (context.Context, chan<- *configpb.TestGroup, func() []*configpb.TestGroup) {
+				ch := make(chan *configpb.TestGroup)
+				var wg sync.WaitGroup
+				wg.Add(1)
+				var got []*configpb.TestGroup
+				ctx, cancel := context.WithCancel(ctx)
+				go func() {
+					defer wg.Done()
+					for {
+						select {
+						case tg := <-ch:
+							got = append(got, tg)
+							if len(got) == 6 {
+								cancel()
+							}
+						case <-ctx.Done():
+							cancel()
+							return
+						}
+					}
+				}()
+
+				return ctx, ch, func() []*configpb.TestGroup {
+					wg.Wait()
+					return got
+				}
+			},
+			freq: time.Microsecond,
+			want: []*configpb.TestGroup{
+				{
+					Name: "hi",
+				},
+				{
+					Name: "there",
+				},
+				{
+					Name: "hi",
+				},
+				{
+					Name: "there",
+				},
+				{
+					Name: "hi",
+				},
+				{
+					Name: "there",
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			ctx, channel, get := tc.receivers(ctx)
+			if err := tc.q.Send(ctx, channel, tc.freq); err != ctx.Err() {
```

Comment:
From reading the code, I would have expected "multi" and "single" to return no error.

It seems like the second `<-ctx.Done()` check on ll. 183-185 are there so that this returns with the "correct" error instead of no error. Is that necessary? It seems better to just have an `expectError` in your cases and have these two modes terminate differently.

Reply:
1) Yes, both `select` statements are necessary.

The first `select{}` is to check whether the context expired while we were waiting for the lock.

The second `select{}` is to detect when the context expires while we're waiting for someone to read the channel we want to write to.

2) Yes the expectation is single scenarios return nil and multi scenarios return the contextual error.

AKA if `ctx.Err() == nil` (because we popped all the elements and the queue is empty) then this will `if err != nil { t.Error() }`.

When `frequency != 0` then `Send()` runs forever, at least until we cancel the context we sent it. Note the `if len(got) == 2 { cancel() }` lines. In this situation `ctx.Err() != nil` and this will `want := ctx.Err(); err != want { t.Error() }`